Code/GraphMol/SmilesParse/CXSmilesOps.cpp

```diff
@@ -0,0 +1,280 @@
+//
+//  Copyright (C) 2016 Greg Landrum
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <RDGeneral/BoostStartInclude.h>
+#include <boost/algorithm/string.hpp>
+#include <boost/foreach.hpp>
+#include <boost/lexical_cast.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/RDKitQueries.h>
+#include <iostream>
+#include "SmilesParse.h"
+#include "SmilesParseOps.h"
+
+namespace SmilesParseOps {
+using namespace RDKit;
+
+namespace parser {
+template <typename Iterator>
+std::string read_text_to(Iterator &first, Iterator last, const char sep,
+                         const char blockend) {
+  Iterator start = first;
+  // EFF: there are certainly faster ways to do this
+  while (first != last && *first != sep && *first != blockend) {
+    ++first;
+  }
+  std::string res(start, first);
+  return res;
+}
+
+template <typename Iterator>
+bool parse_atom_labels(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  if (first >= last || *first != '$') return false;
+  ++first;
+  unsigned int atIdx = 0;
+  while (first != last && *first != '$') {
+    std::string tkn = read_text_to(first, last, ';', '$');
+    if (tkn != "") {
+      mol.getAtomWithIdx(atIdx)->setProp(RDKit::common_properties::atomLabel,
+                                         tkn);
+    }
+    ++atIdx;
+    if (first != last && *first != '$') ++first;
+  }
+  if (first == last || *first != '$') return false;
+  ++first;
+  return true;
+}
+
+template <typename Iterator>
+bool parse_coords(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  if (first >= last || *first != '(') return false;
+
+  RDKit::Conformer *conf = new Conformer(mol.getNumAtoms());
+  mol.addConformer(conf);
+  ++first;
+  unsigned int atIdx = 0;
+  while (first != last && *first != ')') {
+    RDGeom::Point3D pt;
+    std::string tkn = read_text_to(first, last, ';', ')');
+    if (tkn != "") {
+      std::vector<std::string> tokens;
+      boost::split(tokens, tkn, boost::is_any_of(std::string(",")));
+      if (tokens.size() >= 1 && tokens[0].size())
+        pt.x = boost::lexical_cast<double>(tokens[0]);
```

Comment:
I suppose I would like some exception catching here for parse errors.  I feel like I should make some explicitly broken CX smiles for testing